### PR TITLE
During expression equality, check for new ordering information

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

## Rationale for this change
See [issue #9812](https://github.com/apache/datafusion/issues/9812) for rationale.

## What changes are included in this PR?
When an equality expression is added to the `EquivalenceProperties`. In light of this information, we can deduce new information about ordering and constantness. This PR adds this support. 

With the [PR #9813](https://github.com/apache/datafusion/pull/9813) we have this support, in terms of behaviour. However, current solution doesn't update state and cannot generalize well.

## Are these changes tested?
There is an additional TestCase, that showcases increased coverage with this implementation (That test case fails in main branch).

### CODE REVIEW

**Comment by mustafasrepo** [comment]
@suremarc, I would appreciate your review if you have time

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @mustafasrepo 

cc @suremarc -- can you also give this a review if you have a chance?

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 1]
```diff
@@ -198,6 +198,61 @@ impl EquivalenceProperties {
...
+            }
+        } else if self.is_expr_constant(right) {
+            // Right expression is constant, add left as constant
+            if !physical_exprs_contains(&self.constants, left) {
+                self.constants.push(left.clone());
+            }
+        }
+
+        // Discover new valid orderings in the light of new equality
+        // See issue: https://github.com/apache/datafusion/issues/9812 for rationale
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence/properties.rs:215]:
```suggestion
        // Discover new valid orderings in light of the new equality. For a discussion, see:
        // https://github.com/apache/datafusion/issues/9812
```

File: datafusion/physical-expr/src/equivalence/mod.rs [hunk 2]
```diff
@@ -47,48 +46,8 @@ pub fn collapse_lex_req(input: LexRequirement) -> LexRequirement {
...
-            {
-                // Remove the redundant sort
-                return None;
-            }
-
-            Some(item)
-        })
-        .cloned()
-        .collect::<Vec<_>>()
+    // collapse_monotonic_lex_req(output)
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence/mod.rs:49]:
```suggestion
```

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 3]
```diff
@@ -198,6 +198,61 @@ impl EquivalenceProperties {
         left: &Arc<dyn PhysicalExpr>,
         right: &Arc<dyn PhysicalExpr>,
     ) {
+        // Discover new constants in the light of new ordering
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence/properties.rs:201]:
```suggestion
        // Discover new constants in light of new the equality:
```

**Review by ozankabak — APPROVED** [review]
LGTM, left three minor comments

**Comment by ozankabak** [comment]
Given this is a fairly small piece of code, two reviewers already took a look and all tests pass, I will go ahead and merge it. If there is any issue we discover later on, or @suremarc points out something we are missing, we will address with a follow-on PR. Thanks all

**Comment by suremarc** [comment]
Sorry @alamb @mustafasrepo, I completely missed this. My GH notifications tend to get drowned out by stuff from work, but I've discovered my personal email doesn't have this issue. I will check it every day from now on. 

As for this PR, looks great and I'm glad we don't have to declare equivalences backwards (`f(a) = b`) for the sort optimizations to work. I do think the `collapse_monotonic_lex_req` was a bit of a smell, so it's nice to move that logic closer into the `EquivalenceProperties` code 👍

**Pull request merged**